staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -231,90 +264,115 @@ func (c *celAdmissionController) Validate(
 			// be a bindingInfo for it
 			bindingInfo := c.bindingInfos[namespacedBindingName]
 			binding := bindingInfo.lastReconciledValue
-			if !binding.Matches(a) {
+			matches, err := c.validatorCompiler.BindingMatches(a, o, binding)
+			if err != nil {
+				return err
+			}
+			if !matches {
 				continue
 			}
 
 			var param *unstructured.Unstructured
 
-			// If definition has no paramsource, always provide nil params to
-			// evaluator. If binding specifies a params to use they are ignored.
-			// Done this way so you can configure params before definition is ready.
-			if paramSource := definition.GetParamSource(); paramSource != nil {
-				paramsNamespace, paramsName := binding.GetTargetParams()
+			// If definition has no paramKind, paramRef must be unset in binding.
+			// If definition has paramKind, paramRef is required in binding.
```

Comment:
continuing the thread from https://github.com/kubernetes/kubernetes/pull/113314#discussion_r1013569320

this seems like it make it harder than it needs to be (or maybe impossible, if I'm reading this correctly) to roll out a "no params → using params" or "using params → no params" transition to a policy

> // If definition has no paramKind, paramRef must be unset in binding.

This means if you have a policy currently using paramKind and you want to stop, you have to follow three steps:
1. drop `params` from the policy CEL expressions
2. sweep all policy bindings and clear paramRef
3. drop `paramKind` from the policy

Ordering 1 and 3 is hard when working with declarative flows. Being able to do a single update to the policy would be really nice. For that to not error, we would have to tolerate a binding with a paramRef to a policy with no paramKind. When evaluating, we just wouldn't resolve the params since the policy is not expecting them.

> // If definition has paramKind, paramRef is required in binding.

Don't we have to tolerate a mismatch in at least one direction in order to transition a policy from not using params to using params or vice-versa without deleting all bindings first?

Reply:
Thanks! Scheduled a sync on Monday to discuss.

Reply:
resolved that if both paramKind and paramRef are non-nil, we'll look up the referenced object and treat a lookup failure as an error

if either paramKind or paramRef are nil, we'll treat `params` as nil

Reply:
Note: We reach agreement on passing null params if paramKind or paramRef is unset. Code update and API doc update are pushed.